### PR TITLE
[WIP/PoC] KIALI-1650 Delete Istio-config objects

### PR DESCRIPTION
**Pull request by pilhuhn**
** Describe the change **

Provide an API endpoint that can be used to delete Istio-config objects

If the code is fine, merge it. Otherwise someone else can use it as a blueprint and implement a better solution.

### CODE REVIEW

**Comment by rhqci** [comment]
:heavy_check_mark: Jenkins CI: kiali-core-pr-e2e-test [#126](https://paste.fedoraproject.org/paste/iH9xSYcRuud8BZmgMv3MCg)
  * :heavy_check_mark: Istio-and-Kiali-on-Openshift [#[1470]](https://paste.fedoraproject.org/paste/NlDzdVihRfbkWFRT~WqJBg)
  * :heavy_check_mark: Istio-Upstream [#[336]](https://paste.fedoraproject.org/paste/VhswINdW3~8TESIvfLsryQ)
  * :heavy_check_mark: Kiali-Istio-Checker [#[1489, 1490]](https://paste.fedoraproject.org/paste/YnnJokcAwTvKd3G6LDcuOg)
  * :heavy_check_mark: install-bookinfo [#[840]](https://paste.fedoraproject.org/paste/rbVx76q6WYSKS2cguQ-3tA)
  * :heavy_check_mark: Kiali-on-Openshift [#[840]](https://paste.fedoraproject.org/paste/Uxud5jNqUf6q-rjcZWFdaw)
  * :heavy_check_mark: run-kiali-e2e-tests [#[273]](https://paste.fedoraproject.org/paste/cRNjVdIk~qorxKDhHMOTqg)


KIALI_NAME=kialicorepr533-istio-system

**Comment by jmazzitelli** [comment]
This is a big change to Kiali because it means we will no longer be read-only. As of now, Kiali is a read-only app (that is, it does not add, remove, or modify anything in OpenShift or Istio). Because this PR can now destroy things, it introduces a paradigm shift in Kiali - we are now a read-write system. So, we need to be careful here - what kind of (additional?) security do we need around this? Is our simple authentication good enough? What kinds of "bad" things can happen if a hacker accesses this API?

**Comment by abonas** [comment]
is there a need in any cascading/related deletion of objects? I don't have any good example for Istio, so perhaps it's not an issue, but the general use case is "if I delete object of type A, then object of type B has no point to exist and should be deleted as well", or "if I delete object of type A, then its reference has to be deleted from object B"

**Comment by jotak** [comment]
FYI i starting working on this ticket (KIALI-1650), based on this PR + adding other stuff.
One question I have, is if we want to provide new openshift/kube templates for deploying kiali readonly ? Or maybe that can be parametrized via helm charts (not sure for openshift) ?

File: handlers/istio_config.go [hunk 1]
```diff
@@ -121,6 +121,37 @@ func IstioConfigDetails(w http.ResponseWriter, r *http.Request) {
 	RespondWithJSON(w, http.StatusOK, istioConfigDetails)
 }
 
+func IstioConfigDelete(w http.ResponseWriter, r* http.Request) {
```
[the remark this answers is not in the document]
Comment (jotak) [reply handlers/istio_config.go]: I think there's a difference in naming between the lower layer (kubernetes client) where we talk more about "istio object" in a general way, and in the upper layer where it's "Istio Config".

In low level, probably we wanted a generic word to encompass both objects under apis "config.istio.io" (ex: rules, quotaspecs) and "networking.istio.io" (ex: destinationrules, virtualservices).

But on the upper level we were probably more guided by the UI, where istio objects are all gathered under the page named "Istio Config" (and so does the menu item)

Who said naming was difficult? :)

File: kubernetes/istio_details_service.go [hunk 2]
```diff
@@ -45,6 +46,15 @@ func (in *IstioClient) GetIstioDetails(namespace string, serviceName string) (*I
 	return &istioDetails, nil
 }
 
+func (in *IstioClient) DeleteIstioObject(namespace string, objectType string, objectName string) (error) {
+	log.Infof("DeleteIstioObject input: %s %s %s", namespace, objectType, objectName)
+	_, err := in.istioNetworkingApi.Delete().Namespace(namespace).Resource(objectType).Name(objectName).Do().Get()
```
Comment (jotak) [inline kubernetes/istio_details_service.go]: That looks weird: `.delete() ... .get()`, really? :)

Comment (pilhuhn) [reply kubernetes/istio_details_service.go]: I guess I expected the final Get() to retrieve the return code.

If you look around, you'll also see `Api.Get().Namespace(namespace).Resource(virtualServices).Do().Get()` where there are also two `Get()` :-)

File: handlers/istio_config.go [hunk 3]
```diff
@@ -121,6 +121,37 @@ func IstioConfigDetails(w http.ResponseWriter, r *http.Request) {
...
+	params := mux.Vars(r)
+	namespace := params["namespace"]
+	objectType := params["object_type"]
+	object := params["object"]
```
[the remark this answers is not in the document]
Comment (pilhuhn) [reply handlers/istio_config.go]: The name

File: handlers/istio_config.go [hunk 4]
```diff
@@ -121,6 +121,37 @@ func IstioConfigDetails(w http.ResponseWriter, r *http.Request) {
 	RespondWithJSON(w, http.StatusOK, istioConfigDetails)
 }
 
+func IstioConfigDelete(w http.ResponseWriter, r* http.Request) {
+	params := mux.Vars(r)
+	namespace := params["namespace"]
+	objectType := params["object_type"]
```
[the remark this answers is not in the document]
Comment (pilhuhn) [reply handlers/istio_config.go]: The code is more generic than the naming implies and I guess also the rest-url could be made more generic.

File: handlers/istio_config.go [hunk 5]
```diff
@@ -121,6 +121,37 @@ func IstioConfigDetails(w http.ResponseWriter, r *http.Request) {
...
+		return
+	}
+
+
+	err = business.IstioConfig.DeleteIstioConfigDetail(namespace,objectType,object)
+	if err != nil {
+		log.Error(err)
+		RespondWithError(w,http.StatusInternalServerError, err.Error())
+	} else {
+		RespondWithJSON(w, http.StatusOK, "Deleted")
```
[the remark this answers is not in the document]
Comment (pilhuhn) [reply handlers/istio_config.go]: The content is the 'Deleted' message :)  204 would be fine as well. The 200 'Deleted'  makes it more human consumable.

**Comment by pilhuhn** [comment]
@abonas I don't think there is cascading, but I am obviously not sure.
@jotak Not sure if we need r/o templates. But then they may be helpful for some.

@jotak Change things as you see fit.

**Comment by lucasponce** [comment]
Close in favor of #566

**Pull request closed without merging**